Enable adding extra setting tabs

yousifalraheem opened the pull request:
## Hey, I just made a Pull Request!
closes #9805

This PR introduces the ability to add extra setting tabs to the user settings page. By the default it will not constrain how the users want to create the content of these tabs, however, the README of the `@backstage/plugin-user-settings` has been updated to suggest using a similar layout to standardise the look and feel across all setting tabs.

![Screenshot 2022-03-07 at 18 35 54](https://user-images.githubusercontent.com/10633734/157087458-9653fc78-6b05-437e-b1fb-a302025fecec.png)

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

yousifalraheem commented:
@freben .. Please review this PR 🙏

benjdlambert commented:
Thinking that we probably want to have some form of solution that uses `useOutlet` here, so that we can pass in `routes` to be able to link to the different tabs. If you take a look how the `CatalogPage` is provided in the [catalog customization docs](https://backstage.io/docs/features/software-catalog/catalog-customization) you'll see that you can provide routes.

Basically being able to define them as children of the `UserSettings` plugin, and be able to make them renderable.

yousifalraheem commented:
@benjdlambert .. But that means that the only way for users to add a tab to the settings is by actually creating a custom settings page from scratch. The catalog example shows a total replacement of the index page `CustomCatalogPage`.  This is because it is designed to either render the default page or the outlet.
https://github.com/backstage/backstage/blob/master/plugins/catalog/src/components/CatalogPage/CatalogPage.tsx#L27

```tsx
export function CatalogPage(props: DefaultCatalogPageProps) {
  const outlet = useOutlet();

  return outlet || <DefaultCatalogPage {...props} />;
}
```

I think it's much better to for users to add extra tabs by just passing tabs as props and keep everything as is.

As for the links, the original setup already uses `TabbedLayout.Route` component from `@backstage/core-components` which creates links. In this case, I chose to generate the tab path from making the tab title lowercase and kebab-case. I'm assuming that now one will have 2 tabs with the same title, therefore, it will be unique. In the example I added, the title was `Advanced` and the link to the tab would be `/settings/advanced`.

yousifalraheem commented:
@benjdlambert .. I think your suggested approach makes more sense and it is cleaner. I have implemented the outlet approach and now extra tabs are passed like this:

```jsx
<Route path="/settings" element={<UserSettingsPage />}>
  <Route path="/advanced" element={<AdvancedSettings />}>
    Advanced
  </Route>
</Route>
```

I have updated the README to reflect this change.

benjdlambert commented:
Nice! This is looking cleaner! I think it's a little unfortunate the the title is just the children of the `Route` and it's a little misleading as normally that's the content what's going to be rendered. I wonder if it makes more sense to use `TabbedLayout.Route` instead directly which has the title prop already which you can pass in. That way you don't need to do the mapping either. What we could also do, is wrap up these route entries in something that separates it from the other children that may eventually be passed in.

So the API could look something like this:

```tsx
<Route path="/settings" element={<UserSettingsPage />}>
  <UserSettingsExtraTabs>
    <TabbedLayout.Route title="My Extra Content">
      <AdvancedSettings />
    </TabbedLayout.Route>
  </UserSettingsExtraTabs>
</Route>
```

@rugvip thoughts on using the `TabbedLayout.Route` explicitly here rather than wrapping it up in something else?

As for the implementation of the `UserSettingsExtraTabs` it's basically going to be a copy of the `ScaffolderFieldExtensions` component which wraps up any additional extensions in a nice way. You might be able to take advantage of `useElementFilter` to crawl the JSX tree to grab the `UserSettingsExtraTab` element and then the children inside that.

Rugvip commented:
@benjdlambert Yep, agree that using `Route` has the potential to confuse. I think going with something like `UserSettingsExtraTabs` and could perhaps use something like `UserSettingsExtraTabs.Tab` rather than `TabbedLayout.Route`?

yousifalraheem commented:
@benjdlambert @Rugvip .. Can you please help me understand the error in the pipeline?

benjdlambert commented on plugins/user-settings/src/components/UserSettingsTab/UserSettingsTab.tsx:
```diff
@@ -0,0 +1,27 @@
+  title: React.ReactNode;
+}
+
+export const UserSettingsTab: React.FC<Props> = ({ children }) => {
```
We try to avoid using `React.FC` and just use `PropsWithChildren` instead.

benjdlambert reviewed (commented):
Thanks again for the work! Just a few more comments here that I think we need to address.

yousifalraheem commented on plugins/user-settings/api-report.md:
```diff
@@ -108,6 +109,12 @@ export const UserSettingsProfileCard: () => JSX.Element;
 // @public (undocumented)
 export const UserSettingsSignInAvatar: ({ size }: Props_5) => JSX.Element;
 
+// Warning: (ae-forgotten-export) The symbol "Props" needs to be exported by the entry point index.d.ts
```
Oh I didn't know warnings are embedded in the generated markdown, I would've assumed to find these warnings in the console when running the command. I'll address them.

yousifalraheem commented on .changeset/three-rivers-remain.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-user-settings': minor
```
changed to `patch`

yousifalraheem commented:
@benjdlambert .. I have added the changes you requested, please re-review 🙂

yousifalraheem commented:
@benjdlambert .. The api-report warnings have been addressed, please re-review

benjdlambert reviewed (approved):
Nice :tada: think this looks good!
